File: backend/app/schemas.py
"""API request/response schemas."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class UserOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    email: str
    name: str | None = None
    avatar_url: str | None = None


class ResearchCreateIn(BaseModel):
    query: str = Field(min_length=3, max_length=2000)


class RunOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    query: str
    status: str
    current_node: str | None = None
    created_at: datetime


class ReviewIn(BaseModel):
    decision: Literal["approve", "edit", "reject"]
    edited_content: str | None = None
    feedback: str | None = None


class CitationOut(BaseModel):
    n: int
    title: str
    url: str


class ReportOut(BaseModel):
    run_id: uuid.UUID
    query: str
    summary: str | None
    content_markdown: str
    citations: list[CitationOut]
    word_count: int
    created_at: datetime


class ReportListItem(BaseModel):
    run_id: uuid.UUID
    query: str
    status: str
    created_at: datetime
    has_report: bool

File: frontend/lib/schemas.ts
import { z } from "zod"

export const userSchema = z.object({
  id: z.string(),
  // Zod 4: top-level z.email() (the old z.string().email() is deprecated).
  email: z.email(),
  name: z.string().nullable().optional(),
  avatar_url: z.string().nullable().optional(),
})

export const runSchema = z.object({
  id: z.string(),
  query: z.string(),
  status: z.string(),
  current_node: z.string().nullable().optional(),
  created_at: z.string(),
})

export const citationSchema = z.object({
  n: z.number(),
  title: z.string(),
  url: z.string(),
})

export const reportSchema = z.object({
  run_id: z.string(),
  query: z.string(),
  summary: z.string().nullable(),
  content_markdown: z.string(),
  citations: z.array(citationSchema),
  word_count: z.number(),
  created_at: z.string(),
})

export const reportListItemSchema = z.object({
  run_id: z.string(),
  query: z.string(),
  status: z.string(),
  created_at: z.string(),
  has_report: z.boolean(),
})

export const reportListSchema = z.array(reportListItemSchema)

// --- SSE event payloads ---

export const hitlPayloadSchema = z.object({
  summary: z.string().default(""),
  draft: z.string().default(""),
  citations: z.array(citationSchema).default([]),
})

export type User = z.infer<typeof userSchema>
export type Run = z.infer<typeof runSchema>
export type Citation = z.infer<typeof citationSchema>
export type Report = z.infer<typeof reportSchema>
export type ReportListItem = z.infer<typeof reportListItemSchema>
export type HitlPayload = z.infer<typeof hitlPayloadSchema>

export type ReviewDecision = "approve" | "edit" | "reject"
